File: app/src/main/java/de/fhswf/fahrradverleih/android/api/remote/RemoteRadApi.java
package de.fhswf.fahrradverleih.android.api.remote;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.Volley;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.JsonReader;
import com.squareup.moshi.JsonWriter;
import com.squareup.moshi.Moshi;
import com.squareup.moshi.Types;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.lang.reflect.Type;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

import de.fhswf.fahrradverleih.android.api.OnFailure;
import de.fhswf.fahrradverleih.android.api.OnSuccess;
import de.fhswf.fahrradverleih.android.api.RadApi;
import de.fhswf.fahrradverleih.android.api.RadApiException;
import de.fhswf.fahrradverleih.android.model.Ausleihe;
import de.fhswf.fahrradverleih.android.model.Benutzer;
import de.fhswf.fahrradverleih.android.model.Fahrrad;
import de.fhswf.fahrradverleih.android.model.LoginResult;
import de.fhswf.fahrradverleih.android.model.Station;
import de.fhswf.fahrradverleih.android.util.Prefs;

public class RemoteRadApi implements RadApi {
    // TODO: Base-URL aus .env lesen
    private static final String BASE_URL = "http://10.0.0.50:3000/api";

    // Wrapper für kürzeren Zugriff auf Methoden
    private static final int POST = Request.Method.POST;
    private static final int GET = Request.Method.GET;

    @NonNull
    private final Context context;

    @Nullable
    private String token;

    private final Moshi moshi;
    private final RequestQueue requestQueue;

    public RemoteRadApi(@NonNull Context context) {
        this.context = context;

        // Token aus Prefs lesen
        this.token = Prefs.getString(context, Prefs.KEY_TOKEN);

        // JSON Serialisierung vorbereiten
        this.moshi = new Moshi.Builder()
                .add(LocalDateTime.class, new DateTimeAdapter())
                .build();

        // Volley Request-Queue
        this.requestQueue = Volley.newRequestQueue(context);
    }

    @Override
    public void login(@NonNull String email, @NonNull String secret,
                      @NonNull OnSuccess<LoginResult> onSuccess,
                      @NonNull OnFailure onFailure) {
        try {
            post(LoginResult.class, "/benutzer/login",
                    new JSONObject()
                            .put("email", email)
                            .put("secret", secret),
                    onSuccess, onFailure);
        } catch (JSONException ignored) {
        }
    }

    @Override
    public void auth(@NonNull OnSuccess<Void> onSuccess,
                     @NonNull OnFailure onFailure) {
        get(Void.class, "/benutzer/auth", new JSONObject(), onSuccess, onFailure);
    }

    @Override
    public void getBenutzer(@NonNull OnSuccess<Benutzer> onSuccess,
                            @NonNull OnFailure onFailure) {
        get(Benutzer.class, "/benutzer/details", new JSONObject(), onSuccess, onFailure);
    }

    @Override
    public void setBenutzer(@NonNull Benutzer benutzer,
                            @NonNull OnSuccess<Benutzer> onSuccess,
                            @NonNull OnFailure onFailure) {
        try {
            post(Benutzer.class, "/benutzer/details",
                    new JSONObject(moshi.adapter(Benutzer.class).toJson(benutzer)),
                    onSuccess, onFailure);
        } catch (JSONException ignored) {
        }
    }

    @Override
    public void getAusleihen(@NonNull OnSuccess<List<Ausleihe>> onSuccess,
                             @NonNull OnFailure onFailure) {
        get(Types.newParameterizedType(List.class, Ausleihe.class),
                "/benutzer/ausleihen", new JSONObject(), onSuccess, onFailure);
    }

    @Override
    public void neueAusleihe(@NonNull String radId, @NonNull LocalDateTime von,
                             @NonNull LocalDateTime bis,
                             @NonNull OnSuccess<Ausleihe> onSuccess,
                             @NonNull OnFailure onFailure) {
        try {
            post(Ausleihe.class, "benutzer/ausleihen/neu",
                    new JSONObject()
                            .put("fahrrad", new JSONObject().put("id", radId))
                            .put("von", von.format(DateTimeFormatter.ISO_DATE_TIME))
                            .put("bis", bis.format(DateTimeFormatter.ISO_DATE_TIME)),
                    onSuccess, onFailure);
        } catch (JSONException ignored) {
        }
    }

    @Override
    public void beendeAusleihe(@NonNull String ausleiheId, @NonNull String stationId,
                               @NonNull OnSuccess<Ausleihe> onSuccess,
                               @NonNull OnFailure onFailure) {
        try {
            post(Ausleihe.class, "/benutzer/ausleihen/ende",
                    new JSONObject()
                            .put("ausleihe", new JSONObject().put("id", ausleiheId))
                            .put("station", new JSONObject().put("id", stationId))
            , onSuccess, onFailure);
        } catch (JSONException ignored) {
        }
    }

    @Override
    public void getStationen(@NonNull OnSuccess<List<Station>> onSuccess,
                             @NonNull OnFailure onFailure) {
        get(Types.newParameterizedType(List.class, Station.class), "/stationen",
                new JSONObject(), onSuccess, onFailure);
    }

    @Override
    public void getRaeder(@NonNull String stationId,
                          @NonNull OnSuccess<List<Fahrrad>> onSuccess,
                          @NonNull OnFailure onFailure) {
        get(Types.newParameterizedType(List.class, Fahrrad.class),
                String.format("/stationen/%s/raeder", stationId),
                new JSONObject(), onSuccess, onFailure);
    }

    /**
     * Aktualisiert das verwendete Token und speichert es mit {@link Prefs}.
     *
     * @param token Neues Token. <code>null</code> löscht das gespeicherte Token.
     */
    public void updateToken(@Nullable String token) {
        this.token = token;

        if (token != null) Prefs.setString(context, Prefs.KEY_TOKEN, token);
        else Prefs.remove(context, Prefs.KEY_TOKEN);
    }

    private <T> void get(@NonNull Type type,
                         @NonNull String path, @NonNull JSONObject args,
                         @NonNull OnSuccess<T> onSuccess,
                         @NonNull OnFailure onFailure) {
        requestQueue.add(new RadRequest(
                GET, BASE_URL + path, token, args,
                response -> process(type, response, onSuccess, onFailure),
                new RadApiFailureListener(onFailure)
        ));
    }

    private <T> void post(@NonNull Type type,
                          @NonNull String path, @NonNull JSONObject args,
                          @NonNull OnSuccess<T> onSuccess,
                          @NonNull OnFailure onFailure) {
        requestQueue.add(new RadRequest(
                POST, BASE_URL + path, token, args,
                response -> process(type, response, onSuccess, onFailure),
                new RadApiFailureListener(onFailure)
        ));
    }

    private <T> void process(@NonNull Type type, String in,
                             @NonNull OnSuccess<T> onSuccess,
                             @NonNull OnFailure onFailure) {
        try {
            if(type == Void.class) {
                onSuccess.onSuccess(null);
            } else {
                onSuccess.onSuccess(moshi.<T>adapter(type).fromJson(in));
            }
        } catch (Exception e) {
            onFailure.onFailure(new RadApiException("Ungültige Antwort.", e));
        }
    }

    /**
     * Adapter für {@link Moshi}, um Timestamps die in String-Form vorliegen
     * in Objekte zu konvertieren.
     */
    static class DateTimeAdapter extends JsonAdapter<LocalDateTime> {

        @Override
        public LocalDateTime fromJson(JsonReader reader) throws IOException {
            if(reader.peek() == null) return reader.nextNull();
            String value = reader.nextString();
            return LocalDateTime.parse(value, DateTimeFormatter.ISO_DATE_TIME);
        }

        @Override
        public void toJson(@NonNull JsonWriter writer, LocalDateTime value) throws IOException {
            if(value == null) writer.nullValue();
            else writer.value(value.format(DateTimeFormatter.ISO_DATE_TIME));
        }
    }
}

File: app/src/main/java/de/fhswf/fahrradverleih/android/api/remote/RemoteRadApi.kt
package de.fhswf.fahrradverleih.android.api.remote

import android.content.Context
import com.android.volley.Request
import com.android.volley.RequestQueue
import com.android.volley.toolbox.Volley
import de.fhswf.fahrradverleih.android.api.RadApi
import de.fhswf.fahrradverleih.android.api.RadApiException
import de.fhswf.fahrradverleih.android.model.*
import de.fhswf.fahrradverleih.android.util.Prefs
import kotlinx.serialization.SerializationException
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.decodeFromJsonElement
import org.json.JSONObject
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * Live-Implementierung der {@link RadApi}.
 *
 * Verwendet Volley für HTTP-Requests.
 */
class RemoteRadApi(
    private val context: Context,
    private val requestQueue: RequestQueue = Volley.newRequestQueue(context),
) : RadApi {

    // TODO: Base-URL aus Environment lesen
    private val baseUrl: String = "http://10.0.0.50:3000/api"
    private var token: String = Prefs.getString(context, Prefs.KEY_TOKEN, "")!!

    override fun login(
        email: String,
        secret: String,
        onSuccess: (result: LoginResult) -> Unit,
        onFailure: (e: RadApiException) -> Unit
    ) {
        requestQueue.add(RadRequest(
            Request.Method.POST, "$baseUrl/benutzer/login", token,
            JSONObject().put("email", email).put("secret", secret),
            { response -> onSuccess(Json.decodeFromString(response)) },
            { error -> onFailure(RadApiException(error.message, error)) }
        ))
    }

    override fun auth(onSuccess: () -> Unit, onFailure: (e: RadApiException) -> Unit) {
        requestQueue.add(RadRequest(
            Request.Method.GET, "$baseUrl/benutzer/auth", token, null,
            { onSuccess() },
            { e -> onFailure(RadApiException(e.message, e)) }
        ))
    }

    override fun getBenutzer(
        onSuccess: (result: Benutzer) -> Unit,
        onFailure: (e: RadApiException) -> Unit
    ) {
        requestQueue.add(RadRequest(
            Request.Method.GET, "$baseUrl/benutzer/details", token, null,
            { response -> onSuccess(Json.decodeFromString(response)) },
            { error -> onFailure(RadApiException(error.message, error)) }
        ))
    }

    override fun setBenutzer(
        benutzer: Benutzer,
        onSuccess: (result: Benutzer) -> Unit,
        onFailure: (e: RadApiException) -> Unit
    ) {
        requestQueue.add(RadRequest(
            Request.Method.POST, "$baseUrl/benutzer/details", token,
            JSONObject(Json.encodeToString(benutzer)),
            { response -> onSuccess(Json.decodeFromString(response)) },
            { error -> onFailure(RadApiException(error.message, error)) }
        ))
    }

    override fun getAusleihen(
        onSuccess: (result: List<Ausleihe>) -> Unit,
        onFailure: (e: RadApiException) -> Unit
    ) {
        requestQueue.add(RadRequest(
            Request.Method.GET, "$baseUrl/benutzer/ausleihen", token, null,
            { response -> onSuccess(parseList(response)) },
            { error -> onFailure(RadApiException(error.message, error)) }
        ))
    }

    override fun neueAusleihe(
        radId: String,
        von: LocalDateTime,
        bis: LocalDateTime,
        onSuccess: (result: Ausleihe) -> Unit,
        onFailure: (e: RadApiException) -> Unit
    ) {
        requestQueue.add(RadRequest(
            Request.Method.POST, "$baseUrl/benutzer/ausleihen/neu", token,
            JSONObject()
                .put("fahrrad", JSONObject().put("id", radId))
                .put("von", von.format(DateTimeFormatter.ISO_DATE_TIME))
                .put("bis", bis.format(DateTimeFormatter.ISO_DATE_TIME)),
            { response -> onSuccess(Json.decodeFromString(response)) },
            { error -> onFailure(RadApiException(error.message, error)) }
        ))
    }

    override fun beendeAusleihe(
        ausleiheId: String,
        stationId: String,
        onSuccess: (result: Ausleihe) -> Unit,
        onFailure: (e: RadApiException) -> Unit
    ) {
        requestQueue.add(RadRequest(
            Request.Method.POST, "$baseUrl/benutzer/details", token,
            JSONObject()
                .put("ausleihe", JSONObject().put("id", ausleiheId))
                .put("station", JSONObject().put("id", stationId)),
            { response -> onSuccess(Json.decodeFromString(response)) },
            { error -> onFailure(RadApiException(error.message, error)) }
        ))
    }

    override fun getStationen(
        onSuccess: (result: List<Station>) -> Unit,
        onFailure: (e: RadApiException) -> Unit
    ) {
        requestQueue.add(RadRequest(
            Request.Method.GET, "$baseUrl/stationen", token, null,
            { response -> onSuccess(parseList(response)) },
            { error -> onFailure(RadApiException(error.message, error)) }
        ))
    }

    override fun getRaeder(
        stationId: String,
        onSuccess: (result: List<Fahrrad>) -> Unit,
        onFailure: (e: RadApiException) -> Unit
    ) {
        requestQueue.add(RadRequest(
            Request.Method.GET, "$baseUrl/stationen/$stationId/raeder", token, null,
            { response -> onSuccess(parseList(response)) },
            { error -> onFailure(RadApiException(error.message, error)) }
        ))
    }

    /**
     * Verarbeitet ein JSON-Array von Elementen T, welches ursprünglich als String vorliegt.
     *
     * Setzt für die Deserialisierung voraus, dass T {@link kotlinx.serialization.Serializable} ist.
     */
    private inline fun <reified T> parseList(input: String): List<T> {
        try {
            // Parsen des JSON-Arrays
            val jsonArray = Json.parseToJsonElement(input)
            if (jsonArray !is JsonArray) throw SerializationException("Unexpected type.")

            // Liste bauen
            val list = ArrayList<T>(jsonArray.size)
            for (c in jsonArray) list.add(Json.decodeFromJsonElement(c))
            return list
        } catch (e: Exception) {
            throw RadApiException("Verarbeiten der Antwort fehlgeschlagen.", e)
        }
    }

    fun updateToken(token: String) {
        // Token speichern
        Prefs.setString(context, Prefs.KEY_TOKEN, token)

        // In Instanz anwenden
        this.token = token
    }
}